secret_store/src/listener/service_contract.rs

```diff
@@ -0,0 +1,314 @@
+					self.self_key_pair.public());
+				return Box::new(::std::iter::empty());
+			},
+		};
+
+		let contract = self.contract.read();
+		let length = match contract.address == Default::default() {
+			true => 0.into(),
+			false => {
+				let do_call = |a, d| future::done(client.call_contract(BlockId::Latest, a, d));
+				contract.server_key_generation_requests_count(&do_call).wait()
+					.map_err(|error| {
+						warn!(target: "secretstore", "{}: call to server_key_generation_requests_count failed: {}",
+							self.self_key_pair.public(), error);
+						error
+					})
+					.unwrap_or_default()
+			},
+		};
+
+		Box::new(PendingRequestsIterator {
+			client: client,
+			contract: contract.clone(),
+			self_key_pair: self.self_key_pair.clone(),
+			index: 0.into(),
+			length: length,
+		})
+	}
+
+	fn publish_server_key(&self, server_key_id: &ServerKeyId, server_key: &Public) -> Result<(), String> {
+		// only publish if contract address is set && client is online
+		let contract = self.contract.read();
+		if contract.address == Default::default() {
+			// it is not an error, because key could be generated even without contract
+			return Ok(());
+		}
+		let client = match self.client.upgrade() {
+			Some(client) => client,
+			None => return Err("client is required to publish key".into()),
+		};
+
+		// only publish key if contract waits for publication
+		// failing is ok here - it could be that enough confirmations have been recevied
+		// or key has been requested using HTTP API
+		let do_call = |a, d| future::done(client.call_contract(BlockId::Latest, a, d));
+		let self_address = public_to_address(self.self_key_pair.public());
+		if contract.get_server_key_confirmation_status(&do_call, server_key_id.clone(), self_address).wait().unwrap_or(false) {
+			return Ok(());
+		}
+
+		// prepare transaction data
+		let server_key_hash = keccak(server_key);
+		let signed_server_key = self.self_key_pair.sign(&server_key_hash).map_err(|e| format!("{}", e))?;
+		let signed_server_key: Signature = signed_server_key.into_electrum().into();
+		let transaction_data = contract.encode_server_key_generated_input(server_key_id.clone(),
+			server_key.to_vec(),
+			signed_server_key.v(),
+			signed_server_key.r().into(),
+			signed_server_key.s().into()
+		)?;
+
+		// send transaction
+		if contract.address != Default::default() {
+			client.transact_contract(
+				contract.address.clone(),
+				transaction_data
+			).map_err(|e| format!("{}", e))?;
+		}
+
+		Ok(())
+	}
+}
+
+impl Iterator for PendingRequestsIterator {
+	type Item = (bool, ServiceTask);
+
+	fn next(&mut self) -> Option<(bool, ServiceTask)> {
+		if self.index >= self.length {
+			return None;
+		}
+
+		let index = self.index.clone();
+		self.index = self.index + 1.into();
+
+		let self_address = public_to_address(self.self_key_pair.public());
+		let do_call = |a, d| future::done(self.client.call_contract(BlockId::Latest, a, d));
```

Comment:
Note that you are using `Latest` here and the state may change in the meantime. Are we sure that the contract content does not change (including `length`)?

Reply:
After reviewing the rest of the code I think we should wait for some confirmations - the latest block can end up on a fork (even if we trust authorities) and the transaction may be included a bit later, but the state can be different.

I think for security we should fetch `latest - 3` and refer to it using hash, not number.

Reply:
Right. Initial idea behind this listener was to react as soon as possible. The only bad thing that could happen is that SS will generate the key for free (without fee paid by requester && without publishing it on the contract). So I considered this as a possible annoyance, not an issue.

This is completely different from next PR (auto-migration), where reaction to 'forked' changes could potentially lead to data loss. That's why there were no any confirmations requirements && in next PR they're stronger than 3 confirmations.

But I guess you're right here and it is an actual issue. So I've updated PR:
1) added required confirmations to `PendingRequestsIterator`
2) since we wait for several blocks anyway, I've completely removed reaction to contract' events - I suppose adding support for this will increase complexity of the code significantly without any actual output (we won't react to events immediately anyway)

Reply:
Me-of-today disagree with me-of-yesterday - think that both retry && reactions to contract events are important. And that is why:
1) retry was designed to deal with stuck requests - e.g. which have been received when SS wasn't fully-connected. That's why it executes rarely && stops processing stuck requests just after first `MAX_FAILED_RETRY_REQUESTS` fail (currently 1)
2) reacting to contract events was designed to respond to requests as soon as possible (in my setup key was generated within <1s && published in the next block). Now this guarantee has shifted (we wait for 3 blocks before processing request), but we still want to process:
2.1) all incoming requests
2.2) requests that have exact age of 3 blocks (omitting reorg case here) - i.e. as soon as possible, given 3 confirmations required

So I've reverted removal and instead of reacting to passed `enacted` blocks (in `read_logs`) I keep track of last-block-from-which-events-were-read-from && use the `Client::tree_route` to build route from this block to the latest block with 3 confirmations.